### ISSUE TITLE
r748, Texplorer, Regenerate thumb problems

### ISSUE THREAD

**Issue by ghost** [ME3Explorer/ME3Explorer#299]
I'm not sure if this is related to #296 or not so I'm starting another issue just to be safe.

Regenerate Thumb in Texplorer is doing something strange, and I'm guessing it's related to alpha. 

There's two specs in the screenshot below. After building a tree with the current rev I did a regenerate thumb on the lower one and it changed it's appearance to mostly black (what you see). The texture looks identical if I double-click the thumb to see the largest mip. 

![spec_thumb](https://cloud.githubusercontent.com/assets/13645804/14365230/70c48f40-fcc1-11e5-9bb3-cb4edb8fb8be.PNG)

Then, after taking a screenshot I tried the same thing on the top right Spec. The exact same thing happens.

What's particularly strange is that everything looks fine... initially.

**Comment by ghost** [comment]
More info on this.

Regen thumb on PCC-stored GUI textures makes them go blank (See EDI CDX):

![regenthumbb](https://cloud.githubusercontent.com/assets/13645804/14365998/80115344-fcc5-11e5-94fe-b3e3ea057ffa.PNG)

FYI, I just did another fresh treebuild, just in case.

Will check a few more and edit this post as necessary.

---

EDIT: What it seems to be is any GUI-texture that's PCC-stored, and most DXT5 textures with a lot of transparency. Here's a couple more examples:

![regenthumbc](https://cloud.githubusercontent.com/assets/13645804/14366373/428a0ea6-fcc7-11e5-9a92-20e17bfacd4b.PNG)
![regenthumbd](https://cloud.githubusercontent.com/assets/13645804/14366377/44915966-fcc7-11e5-9f0c-f34a60fca953.PNG)
In the last example above, the thumb actually doesn't change, but the texture appearance obviously changes drastically when enlarged.

---

What I'm not clear on is _how_ the DXT5s should look. As in, what's the intent? Should it look like the thumb before or after the regenerate thumb function is used?

Obviously, the disappearance of the PCC-stored textures is a new bug, though.

**Comment by KFreon** [comment]
I am so tired of this. Why can't everyone have the a cbd windows? And why is the toolset still in winforms?
I'll look into it.

**Comment by ghost** [comment]
Sorry :(

For what it's worth, I'm actually seeing more consistency in how the alpha is rendered now. Why it decides to change when the thumbnail is regenerated, though, I don't get.

And, what makes even less sense is how GUI textures are simply lost now upon regen. That might actually be a different issue; not sure.

What's cbd windows, btw?

Also, like I said, I'm not sure what the _intent_ is right now. It could be that I'm not finally seeing what's correct; I don't know. Should the asari head be whitish or blackish?

**Comment by KFreon** [comment]
There's about 10 different GetImage and GetThumbnail functions spread throughout the toolset. This frustrates me greatly, and I might just have to bite the bullet and consolidate them all.

Also "a cbd" was supposed to be "same". Was on the phone :)
The Asari head may well be blackish now. It might be that it was white-ish before because it was transparent and the background was white.

**Comment by ghost** [comment]
> There's about 10 different GetImage and GetThumbnail functions spread throughout the toolset. This frustrates me greatly, and I might just have to bite the bullet and consolidate them all.

That... would definitely be frustrating.

> Also "a cbd" was supposed to be "same". Was on the phone :)

Ha ha.

> The Asari head may well be blackish now. It might be that it was white-ish before because it was transparent and the background was white.

That's exactly what I was wondering. Having it with the white background is far more practical as far as being able to see the texture detail, so it'll be unfortunate if it has to render black from now on.

I assume this is happening due to the changes you made to alpha? Why, though, is it only occurring upon regenerating the thumbs, and what's the deal with the DXT1 PCC-stored textures totally losing their thumbnail upon regen? I went back to rev747 to re-test both problems and neither are present in the previous rev.

**Comment by KFreon** [comment]
That's the problem with Winforms. If this was WPF, I could have a background colour toggle so the user can switch as they wish.
I don't think I can do that in Winforms, but I'll look because this is now looking like I should revert the black background change.

It would only occur on regenerating thumbs because that's the only place I put it. I thought I called it from tree generation, but I guess I don't, hence the 10 functions problem...

No idea why PCC stored is misbehaving. I'll look at it.

**Comment by ghost** [comment]
I thought Texplorer was built in WPF, though? All the fancy sliding panels and stuff?

**Comment by KFreon** [comment]
You might notice how poor the animation is, and how much it breaks when the scaling is changed?
No, I just saw how cool WPF was and wanted some of that, but it was too painful to make the switch completely, so I just ported some things.

tldr I f***ing wish :P

**Comment by ghost** [comment]
Ahh, I see. :(

**Comment by ghost** [comment]
@KFreon -- I just noticed something new with this problem on the current rev (not the new pre-testbuild).

Check out these caps:
![thumbbuga](https://cloud.githubusercontent.com/assets/13645804/14404178/e78b1748-fe24-11e5-9839-d644d24e8544.PNG)
![thumbbug](https://cloud.githubusercontent.com/assets/13645804/14404179/e9ae3fe6-fe24-11e5-9421-d8db03c1f3a7.PNG)

This does seem to be two different problems. The Alpha problem and the thumb regeneration problem.

As for the thumb regeneration, I definitely didn't notice before that the thumb assigned to the texture became offset for everything after one was replaced. Nor did I see the two missing at the end.

Which leaves two questions:
1. Does your fix for this issue address this (seemingly) larger problem? If not, then I'm going to wait for you to link a new prebuild.
2. This seems like a relatively major bug, so how in the world did something like this happen, lol?

**Comment by KFreon** [comment]
Wow...that I have not noticed.
So when does this happen?

**Comment by ghost** [comment]
Well, whatever was so profoundly messed up with the regeneration and the "empty spots" has officially been de-fuckified.

Also, the thumb alpha thing is better, too. The asari head thumb and enlargement now look the same in Texplorer (white). In Image Engine, the alpha + black background makes it 748 enlargement. AND, the fonts are visible again. Seems like we finally have the best of all worlds.

---

Verified in r749 prebuild.
